### PR TITLE
Replace multi-loaderallocator hash implementation in MethodDescBackpatchInfo

### PR DESCRIPTION
**Pull request by davidwrighton**
The multi-loaderallocator table for MethodDescs as used by MethodDescBackpatchInfo is both complex on shutdown (which is very difficult to test), somewhat memory inefficient, and not generalized for use for other similar needs. This change builds a new hashtable that utilizes the GC to handle lifetime, but is still useable from the portions of our codebase which need to use this data structure (and thus can't be written in managed code) See the comment in CrossLoaderAllocatorHash.h for details on how it works. It is expected that this will be used by this, by work adding support to the typesystem to track derived types, and by upcoming profiler api work (which has driven much of the complex size saving work, as it is expected that data structure will be large).

### CODE REVIEW

**Comment by davidwrighton** [comment]
@davmason I've put this up for review as a WIP to get cross platform testing. I believe its functionally correct, but as its a complex mass of templates currently only tested on msvc, I expect to see new and interesting failures on clang. Let me know what you think of the available api surface, and if it will work for your scenario.

**Comment by davmason** [comment]
@davidwrighton I plan to take a look later today

**Comment by davmason** [comment]
Looks good from the API surface area point of view. I don't see any issues

File: src/vm/crossloaderallocatorhash.inl [hunk 1]
```diff
@@ -0,0 +1,1196 @@
...
+        return true;
+
+    // This is a tricky bit of logic used which detects freed loader allocators lazily
+    // and deletes them from the GCHeapHash while looking up or otherwise walking the hashtable
+    // for any purpose.
+    LAHASHDEPENDENTHASHTRACKERREF value = (LAHASHDEPENDENTHASHTRACKERREF)valueInHeap;
+    if (!value->IsLoaderAllocatorLive())
+    {
+#ifndef DACCESS_COMPILE
+        arr->SetAt(index, gcHeap);
```
Comment (janvorli) [inline src/vm/crossloaderallocatorhash.inl]: It would be nice to add a comment here that the gcHeap stored in the entry represents a freed entry. I know it is mentioned somewhere else, but it would improve the readability.

File: src/vm/crossloaderallocatorhash.h [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    static bool VisitTracker(TKey key, LAHASHDEPENDENTHASHTRACKERREF trackerUnsafe, Visitor &visitor);
+    template <class Visitor>
+    static bool VisitTrackerAllEntries(LAHASHDEPENDENTHASHTRACKERREF trackerUnsafe, Visitor &visitor);
+    template <class Visitor>
+    static bool VisitKeyToTrackerAllEntries(OBJECTREF hashKeyEntryUnsafe, Visitor &visitor);
+    static void DeleteEntryTracker(TKey key, LAHASHDEPENDENTHASHTRACKERREF trackerUnsafe);
+
+private:
+    LoaderAllocator *_loaderAllocator = 0;
+    OBJECTHANDLE LAToDependentTrackerHash = 0;
```
Comment (janvorli) [inline src/vm/crossloaderallocatorhash.h]: A nit - can you please make the member naming convention consistent?

File: src/vm/crossloaderallocatorhash.h [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    template <class Visitor>
+    static bool VisitTracker(TKey key, LAHASHDEPENDENTHASHTRACKERREF trackerUnsafe, Visitor &visitor);
+    template <class Visitor>
+    static bool VisitTrackerAllEntries(LAHASHDEPENDENTHASHTRACKERREF trackerUnsafe, Visitor &visitor);
+    template <class Visitor>
+    static bool VisitKeyToTrackerAllEntries(OBJECTREF hashKeyEntryUnsafe, Visitor &visitor);
+    static void DeleteEntryTracker(TKey key, LAHASHDEPENDENTHASHTRACKERREF trackerUnsafe);
+
+private:
+    LoaderAllocator *_loaderAllocator = 0;
```
Comment (janvorli) [inline src/vm/crossloaderallocatorhash.h]: A nit - _loaderAllocator -> _pLoaderAllocator to make it consistent with the naming of function arguments

File: src/vm/crossloaderallocatorhash.inl [hunk 4]
```diff
@@ -0,0 +1,1196 @@
...
+template <class Visitor>
+/*static*/ bool CrossLoaderAllocatorHash<TRAITS>::VisitKeyToTrackerAllEntries(OBJECTREF hashKeyEntryUnsafe, Visitor &visitor)
+{
+    CONTRACTL
+    {
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    struct _gcStruct
```
Comment (janvorli) [inline src/vm/crossloaderallocatorhash.inl]: A nit - can you please remove the _gcStruct? All other places in the code use anonymous structs for the GC protection purpose

File: src/vm/gcheaphashtable.h [hunk 5]
```diff
@@ -0,0 +1,144 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#ifndef GCHEAPHASHTABLE_H
+#define GCHEAPHASHTABLE_H
+
+class GCHeapHashObject;
+
+template <bool remove_supported>
```
Comment (janvorli) [inline src/vm/gcheaphashtable.h]: A nit - can you please consider using camel case naming convention instead of the underscores for the sake of consistency with the other code.

File: src/vm/gcheaphashtable.h [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+    static const INT32 s_density_factor_numerator = 3;
+    static const INT32 s_density_factor_denominator = 4;
+
+    static const INT32 s_densitywithdeletes_factor_numerator = 7;
+    static const INT32 s_densitywithdeletes_factor_denominator = 8;
+
+    static const INT32 s_minimum_allocation = 7;
+
+    static const bool s_remove_supported = remove_supported;
```
Comment (janvorli) [inline src/vm/gcheaphashtable.h]: It seems this constant is not needed as it is not accessed from the outside, so we can use the template parameter directly.

File: src/vm/loaderallocator.cpp [hunk 7]
```diff
@@ -105,6 +102,23 @@ LoaderAllocator::~LoaderAllocator()
 #endif
 }
 
+#ifndef CROSSGEN_COMPILE
+OBJECTREF LoaderAllocator::GetHandleValue(LOADERHANDLE handle)
```
[the remark this answers is not in the document]
Comment (davidwrighton) [reply src/vm/loaderallocator.cpp]: No, that's a legacy of some other changes that got reverted. I'll move it back.

File: src/vm/crossloaderallocatorhash.inl (hunk 4 above)
Comment (davidwrighton) [reply src/vm/crossloaderallocatorhash.inl]: Removed

File: src/vm/gcheaphashtable.h (hunk 6 above)
Comment (davidwrighton) [reply src/vm/gcheaphashtable.h]: Removed

File: src/vm/crossloaderallocatorhash.inl [hunk 8]
```diff
@@ -0,0 +1,1196 @@
...
+        else
+        {
+            pStartOfValuesData[entriesInArrayTotal - 1] = (TValue)(usedEntries);
+            pStartOfValuesData[entriesInArrayTotal - 2] = (TValue)0;
+        }
+    }
+}
+
+template <class TKey_, class TValue_>
+/*static*/ void NoRemoveDefaultCrossLoaderAllocatorHashTraits<TKey_, TValue_>::AddToValuesInHeapMemory(OBJECTREF &keyValueStore, OBJECTREF &newKeyValueStore, const TKey& key, const TValue& value)
```
[the remark this answers is not in the document]
Comment (davidwrighton) [reply src/vm/crossloaderallocatorhash.inl]: We don't need the complication of two parameters. Returning a bool is just simpler. The reason it was there was to avoid any need to allocate a new GC_PROTECT region in case the target function needs to do a GC triggering operation while holding onto a temporary, but that isn't actually necessary at all. So, I've removed the extra parameter, and converted it to a pointer, which looks much better.

File: src/vm/crossloaderallocatorhash.inl [hunk 9]
```diff
@@ -0,0 +1,1196 @@
...
+        // data structure will require that the key's loader allocator is the same as that of this data structure.
+        _ASSERTE(key->GetLoaderAllocator() == _loaderAllocator);
+
+        gc.keyToTrackersHash = KeyToValuesGCHeapHash((GCHEAPHASHOBJECTREF)ObjectFromHandle(KeyToDependentTrackersHash));
+        INT32 index = gc.keyToTrackersHash.GetValueIndex(&key);
+
+        if (index == -1)
+        {
+            addToKeyValuesHash = true;
+            TRAITS::AddToValuesInHeapMemory(gc.keyValueStore, gc.newKeyValueStore, key, pLoaderAllocatorOfValue == _loaderAllocator ? value : NULL);
```
[the remark this answers is not in the document]
Comment (davidwrighton) [reply src/vm/crossloaderallocatorhash.inl]: Done

File: src/vm/gcheaphashtable.h (hunk 5 above)
Comment (davidwrighton) [reply src/vm/gcheaphashtable.h]: done.

**Comment by davidwrighton** [comment]
@dotnet-bot test this please

**Comment by davidwrighton** [comment]
@davmason @kouvel Could you folks review this? Thanks

File: src/vm/method.cpp [hunk 10]
```diff
@@ -4891,10 +4874,10 @@ void MethodDesc::BackpatchEntryPointSlots(PCODE entryPoint, bool isPrestubEntryP
     _ASSERTE(entryPoint != NULL);
     _ASSERTE(MayHaveEntryPointSlotsToBackpatch());
     _ASSERTE(isPrestubEntryPoint == (entryPoint == GetPrestubEntryPointToBackpatch()));
+    _ASSERTE(MethodDescBackpatchInfoTracker::IsLockedByCurrentThread());
```
Comment (kouvel) [inline src/vm/method.cpp]: Is it possible to keep the lock ordering as before, such that the `MethodDescBackpatchInfoTracker`'s lock is only taken in the paths where recording or backpatching slots is necessary? That would keep the lock usage more contained, would limit the lock scope to as little as necessary, and callers would not have to know too much about the lock.
- This function is called by `MethodDesc::SetCodeEntryPoint`, which has another call from `prestub.cpp` that does not take the lock. It is ok for now, but if the lock ordering has to remain this way, it would be good to add `_ASSERTE(!MayHaveEntryPointSlotsToBackpatch())` before that call
- If the lock ordering has to remain this way, the new places where the lock is taken can pass in the result of `MayHaveEntryPointSlotsToBackpatch()` to the `ConditionalLockHolder` such that the lock is only taken when necessary

File: src/vm/rejit.cpp [hunk 11]
```diff
@@ -613,39 +613,44 @@ HRESULT ReJitManager::UpdateActiveILVersions(
...
-        CodeActivationBatch * pCodeActivationBatch = *iter;
-        CodeVersionManager * pCodeVersionManager = pCodeActivationBatch->m_pCodeVersionManager;
 
-        int cMethodsToActivate = pCodeActivationBatch->m_methodsToActivate.Count();
-        if (cMethodsToActivate == 0)
-        {
-            continue;
-        }
+    {
+        MethodDescBackpatchInfoTracker::ConditionalLockHolder lockHolder;
```
Comment (kouvel) [inline src/vm/rejit.cpp]: If the lock ordering has to remain this way, to limit the lock scope it would be better to move it into `CodeVersionManager::SetActiveILCodeVersions` to just before where the `TableLockHolder` is instantiated.

File: src/vm/codeversion.cpp [hunk 12]
```diff
@@ -2157,6 +2157,8 @@ PCODE CodeVersionManager::PublishVersionableCodeIfNecessary(MethodDesc* pMethodD
             pCode = pMethodDesc->PrepareCode(activeVersion);
         }
 
+        MethodDescBackpatchInfoTracker::ConditionalLockHolder lockHolder;
```
Comment (kouvel) [inline src/vm/codeversion.cpp]: Could move this to just before where the `TableLockHolder` is instantiated, and pass in `pMethodDesc->MayHaveEntryPointSlotsToBackpatch()` as the argument, and similarly in other new places where this lock is taken

File: src/vm/tieredcompilation.cpp [hunk 13]
```diff
@@ -710,6 +711,8 @@ void TieredCompilationManager::ActivateCodeVersion(NativeCodeVersion nativeCodeV
     // code version will activate then.
     ILCodeVersion ilParent;
     HRESULT hr = S_OK;
+    MethodDescBackpatchInfoTracker::ConditionalLockHolder lockHolder;
```
Comment (kouvel) [inline src/vm/tieredcompilation.cpp]: Could move this to just before the two places where the `TableLockHolder` is instantiated to decrease the lock scope

File: src/vm/crossloaderallocatorhash.inl [hunk 14]
```diff
@@ -0,0 +1,1216 @@
...
+
+    DWORD entriesInArrayTotal;
+    DWORD usedEntries = ComputeUsedEntries(pKeyValueStore, &entriesInArrayTotal);
+
+    for (DWORD index = 0; index < usedEntries; ++index)
+    {
+        // Capture pKeyLoc and pStartOfValuesData inside of loop, as we aren't protecting these pointers into the GC heap, so they
+        // are not permitted to live across the call to visitor (in case visitor triggers a GC)
+        TKey* pKeyLoc = (TKey*)((I1ARRAYREF)*pKeyValueStore)->GetDirectPointerToNonObjectElements();
+        TValue* pStartOfValuesData = (TValue*)(((I1ARRAYREF)*pKeyValueStore)->GetDirectPointerToNonObjectElements() + sizeof(TKey));
```
Comment (kouvel) [inline src/vm/crossloaderallocatorhash.inl]: Do these need to use `VolatileLoadWithoutBarrier` to prevent the compiler from hoisting them out of the loop?

File: src/vm/crossloaderallocatorhash.inl [hunk 15]
```diff
@@ -0,0 +1,1216 @@
...
+    }
+    CONTRACTL_END;
+
+    static_assert(sizeof(TKey)==sizeof(TValue), "Assume keys and values are the same size");
+
+    bool updatedKeyValueStore = false;
+
+    if (*pKeyValueStore == NULL)
+    {
+        *pKeyValueStore = AllocatePrimitiveArray(ELEMENT_TYPE_I1, (value == NULL) ? sizeof(TKey) : sizeof(TKey) + sizeof(TValue), FALSE);
```
Comment (kouvel) [inline src/vm/crossloaderallocatorhash.inl]: It is necessary to allocate a key/value store when the value is null?

File: src/vm/crossloaderallocatorhash.inl [hunk 16]
```diff
@@ -0,0 +1,1216 @@
...
+        updatedKeyValueStore = true;
+        TKey* pKeyLoc = (TKey*)((I1ARRAYREF)*pKeyValueStore)->GetDirectPointerToNonObjectElements();
+        *pKeyLoc = key;
+        if (value != NULL)
```
Comment (kouvel) [inline src/vm/crossloaderallocatorhash.inl]: Would it be useful to have an `IsNull()` function on the traits similar to traits for `SHash` instead of comparing to null?

File: src/vm/crossloaderallocatorhash.inl [hunk 17]
```diff
@@ -0,0 +1,1216 @@
...
+                COMPlusThrow(kOverflowException);
+
+            // Allocate the new array.
+            I1ARRAYREF newKeyValueStore = (I1ARRAYREF)AllocatePrimitiveArray(ELEMENT_TYPE_I1, newSize*sizeof(TValue) + sizeof(TKey), FALSE);
+
+            // Since, AllocatePrimitiveArray may have triggered a GC, recapture all data pointers from GC objects
+            void* pStartOfNewArray = newKeyValueStore->GetDirectPointerToNonObjectElements();
+            void* pStartOfOldArray = ((I1ARRAYREF)*pKeyValueStore)->GetDirectPointerToNonObjectElements();
+
+            memcpyNoGCRefs(pStartOfNewArray, pStartOfOldArray, ((I1ARRAYREF)*pKeyValueStore)->GetNumComponents());
```
Comment (kouvel) [inline src/vm/crossloaderallocatorhash.inl]: Looks like memcpyNoGCRefs just calls memcpy with the same arguments. Should the last argument be size in bytes? Does `GetNumComponents()` get the number of elements in the array or size in bytes?

Also it looks like this and other tests for `NULL` on the value imply that the key and value must be pointer-sized types. Maybe add a `static_assert` if there isn't one already, next to the one above.

File: src/vm/object.h [hunk 18]
```diff
@@ -2815,4 +2819,128 @@ typedef REF<ExceptionObject> EXCEPTIONREF;
...
+{
+#ifdef DACCESS_COMPILE
+    friend class ClrDataAccess;
+#endif
+    friend class CheckAsmOffsets;
+    friend class MscorlibBinder;
+
+    public:
+    OBJECTREF _trackerOrTrackerSet;
+    OBJECTREF _laLocalKeyValueStore;
```
Comment (kouvel) [inline src/vm/object.h]: It would be nice to add a comment describing the data structure behind `_trackerOrTrackerSet` and what `_laLocalKeyValueStore` would contain relative to where the entry is referenced from. I'm not sure I fully understood it yet but I imagine the key is not necessarily LA-local and may be from a dependency LA, but the value(s) would be from the LA corresponding to this entry.

File: src/vm/rejit.cpp (hunk 11 above)
Comment (davidwrighton) [reply src/vm/rejit.cpp]: The lock must be outside of any use of EE thread suspension, which takes and holds the ThreadStore lock.

File: src/vm/tieredcompilation.cpp (hunk 13 above)
Comment (davidwrighton) [reply src/vm/tieredcompilation.cpp]: This lock needs to be taken outside of any thread suspension activities, I could move it inside the scope, but it wouldn't be in a different location.

File: src/vm/crossloaderallocatorhash.inl (hunk 17 above)
Comment (davidwrighton) [reply src/vm/crossloaderallocatorhash.inl]: GetNumComponents for an I1Array (array of sbytes) returns the number of elements in the array, which is the same as the number of bytes.

About the NULL checks, the intention is to be able to hold pointer like values that are comparable to NULL, but not necessarily require the data to be stored as a pointer. David Mason and I have discussed making this data structure particularly efficient for his purposes (and there are additional works necessary to pull of that efficiency, but one of them is to use 48bit pointers instead of 64bit pointers. (It turns out all of our platforms that we expect to support in the short or medium term restrict pointer support to only actually supporting a 48 bit address space. I'm not going to write in that support now, but if it turns out its useful, I'd like to leave us free to have 48bit pointers.

File: src/vm/crossloaderallocatorhash.inl (hunk 16 above)
Comment (davidwrighton) [reply src/vm/crossloaderallocatorhash.inl]: Yes. I should do that.

File: src/vm/crossloaderallocatorhash.inl [hunk 19]
```diff
@@ -0,0 +1,1196 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#ifndef CROSSLOADERALLOCATORHASH_INL
+#define CROSSLOADERALLOCATORHASH_INL
+#ifdef CROSSLOADERALLOCATORHASH_H
+#ifndef CROSSGEN_COMPILE
+
+#include "gcheaphashtable.inl"
```
[the remark this answers is not in the document]
Comment (davidwrighton) [reply src/vm/crossloaderallocatorhash.inl]: Yes. The inl representations make use of OBJECTREFs and the various type definitions in object.h and frames.h, which are not included until well after the headers are necessary. So, frames.h will include the inl files.

File: src/vm/crossloaderallocatorhash.inl (hunk 14 above)
Comment (davidwrighton) [reply src/vm/crossloaderallocatorhash.inl]: No, since pKeyValueStore is on the heap (in C++ semantics) if visitor may trigger a GC or poll for a GC, the GC triggering path will take a lock which is considered to be a sufficient instruction to the compiler that it can't hoist the load. If it won't trigger a GC, then the hoist would be valid.

File: src/vm/rejit.cpp (hunk 11 above)
Comment (kouvel) [reply src/vm/rejit.cpp]: Oh is that because of the coop mode requirements?

File: src/vm/crossloaderallocatorhash.inl (hunk 14 above)
Comment (kouvel) [reply src/vm/crossloaderallocatorhash.inl]: Gotcha

**Comment by kouvel** [comment]
From my understanding, the dependent handle in `LAHashDependentHashTrackerObject` causes the object to have the same lifetime as the `LoaderAllocator` without keeping the `LoaderAllocator` alive. So when the `LoaderAllocator` appears to be free, the dependent handle can also be freed.

If correct so far, from that point, does the finalizer of `LAHashDependentHashTrackerObject` need to run before the entry in the `GCHeapHash` appears to be deleted? And it looks like only after the dependent handle is noticed to be free (upon the next add or iterate), the value in the key/value store would be removed.

Wondering if there could be an issue where the `LoaderAllocator` is destroyed but memory from it is still referenced by the value in the key/value store. For example, for a backpatchable slot from a dispatch stub or a MethodTable from a dependent `LoaderAllocator`, is there a possibility that it would try to backpatch the slot after the `LoaderAllocator` is destroyed?

File: src/vm/crossloaderallocatorhash.inl [hunk 20]
```diff
@@ -0,0 +1,1216 @@
...
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    if (m_loaderAllocatorToDependentTrackerHash == NULL)
+    {
+        OBJECTREF laToDependentHandleHashObject = AllocateObject(MscorlibBinder::GetExistingClass(CLASS__GCHEAPHASH));
+        m_loaderAllocatorToDependentTrackerHash = m_pLoaderAllocator->GetDomain()->CreateHandle(laToDependentHandleHashObject);
+        m_pLoaderAllocator->RegisterHandleForCleanup(m_loaderAllocatorToDependentTrackerHash);
```
Comment (kouvel) [inline src/vm/crossloaderallocatorhash.inl]: I see only one call to `AssemblyLoaderAllocator::CleanupHandles` from the `AppDomain` and from the name it seems to be relevant to shutdown. How would these hash tables be freed for collectible `LoaderAllocator`s?

File: src/vm/rejit.cpp (hunk 11 above)
Comment (kouvel) [reply src/vm/rejit.cpp]: Actually I don't see why coop mode would require that, and I still don't understand why the lock has to be taken before EE thread suspension, could you elaborate?

Comment (davidwrighton) [reply src/vm/rejit.cpp]: Its more about the issues around allocating on the GC heap. (Not something that happens in the visitation paths, but since you need to use the same crst to protect the add path, you end up with this problem.)

File: src/vm/crossloaderallocatorhash.inl (hunk 20 above)
Comment (davidwrighton) [reply src/vm/crossloaderallocatorhash.inl]: There are 2 types of GCHandles available for use from within the LoaderAllocator logic. There are LoaderAllocator Handles and GC Handles. LoaderAllocator Handles have lifetime that will last till the end of the managed lifetime of the loaderallocator, but their finalizer may be triggered earlier, as part of general LoaderAllocator finalization. (Its even possible through finalization resurrection to make various things come back to semi-life. However, in this path, we use a GC Handle which is registered for cleanup AFTER the LoaderAllocator is considered destroyed by the managed environment. These handles are cleaned up during a later phase of LoaderAllocator shutdown which happens to be managed by the AppDomain class. (The restriction on these things is that they MUST NOT have any managed pointers to managed objects that can keep a LoaderAllocator alive.) Thus yes, the logic in AppDomain.cpp that calls CleanupHandles is relevant to shutdown, but that shutdown refers to LoaderAllocator shutdown, not AppDomain shutdown.

File: src/vm/crossloaderallocatorhash.inl (hunk 15 above)
Comment (davidwrighton) [reply src/vm/crossloaderallocatorhash.inl]: Yes, this is done when the value isn't contained in the same loaderallocator as the key, so we're in the process of handling the multi-loaderallocator case. We need a place to hold the key.

File: src/vm/codeversion.cpp (hunk 12 above)
Comment (davidwrighton) [reply src/vm/codeversion.cpp]: See other comments about lock order and thread suspension.

File: src/vm/object.h (hunk 18 above)
Comment (davidwrighton) [reply src/vm/object.h]: Comment added.

File: src/vm/crossloaderallocatorhash.inl (hunk 17 above)
Comment (davidwrighton) [reply src/vm/crossloaderallocatorhash.inl]: I've added the IsNull method as suggested in another comment.

File: src/vm/methoddescbackpatchinfo.h [hunk 21]
```diff
@@ -5,29 +5,7 @@
...
-//
-// EntryPointSlots and MethodDescEntryPointSlots
-//   - Collection of slots recorded for backpatching
-//   - There is one instance per MethodDescBackpatchInfo for slots allocated in the MethodDesc's LoaderAllocator
-//   - There is one instance per MethodDesc in MethodDescBackpatchInfoTracker, for slots allocated in LoaderAllocators that are 
-//     dependent on the MethodDesc's LoaderAllocator. The dependent LoaderAllocators are also recorded in the
-//     MethodDescBackPatchInfo associated with the MethodDesc's LoaderAllocator.
-
-typedef SHash<PtrSetSHashTraits<LoaderAllocator *>> LoaderAllocatorSet;
+#include "crossloaderallocatorhash.h"
```
[the remark this answers is not in the document]
Comment (davidwrighton) [reply src/vm/methoddescbackpatchinfo.h]: comments removed.

**Comment by davidwrighton** [comment]
> From my understanding, the dependent handle in LAHashDependentHashTrackerObject causes the object to have the same lifetime as the LoaderAllocator without keeping the LoaderAllocator alive. So when the LoaderAllocator appears to be free, the dependent handle can also be invalidated.
> If correct so far, from that point, does the finalizer of LAHashDependentHashTrackerObject need to run before the entry in the GCHeapHash appears to be deleted? And it looks like only after the dependent handle is noticed to be free (upon the next add or iterate), the value in the key/value store would be removed.
> Wondering if there could be an issue where the LoaderAllocator is destroyed but memory from it is still referenced by the value in the key/value store. For example, for a backpatchable slot from a dispatch stub or a MethodTable from a dependent LoaderAllocator, is there a possibility that it would try to backpatch the slot after the LoaderAllocator or the VirtualCallStubManager is destroyed?

No, this isn't an issue. The dependent handle loses its reference to the secondary when the secondary is actually unreachable (even from within a finalizer). The finalizer actually needs to be delayed until there cannot be an attempt to use the dependent handle to find an answer, hence the use of a GC handle with registered cleanup instead of a LoaderAllocator handle.

File: src/vm/method.cpp (hunk 10 above)
Comment (davidwrighton) [reply src/vm/method.cpp]: Assert added. Use of MayHaveEntryPointSlotsToBackpatch as a flag to the conditional lock holder added for cases that I could analyze.

Comment (davidwrighton) [reply src/vm/method.cpp]: Lock ordering changes are needed, but I've followed you suggestions here.

File: src/vm/tieredcompilation.cpp [hunk 22]
```diff
@@ -718,6 +719,8 @@ void TieredCompilationManager::ActivateCodeVersion(NativeCodeVersion nativeCodeV
...
```
Comment (kouvel) [inline src/vm/tieredcompilation.cpp]: This could also pass in `pMethod->MayHaveEntryPointSlotsToBackpatch()` as a parameter

**Comment by kouvel** [comment]
I see, makes sense

**Comment by davidwrighton** [comment]
@dotnet-bot test Windows_NT x64 Checked CoreFX Tests

**Comment by davidwrighton** [comment]
@dotnet-bot test Windows_NT x64 Checked Innerloop Build and Test
